File: ProjectCS/UI/BaseItemPicker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SessionResultsCS.UI
{
    public abstract class BaseItemPicker
    {
        private IList<string> _items;
        protected int PrintedLines { get; set; }

        public int ItemCount => _items.Count;


        public BaseItemPicker(IEnumerable<string> items)
        {
            _items = items.ToList();
        }

        public int Pick()
        {
            PrintedLines = 0;
            PrintItems();

            int selectedIndex = GetInput();

            ClearOutput();
            Console.WriteLine($"{selectedIndex + 1} - {_items[selectedIndex]}");

            return selectedIndex;
        }


        protected abstract int GetInput();


        private void PrintItems()
        {
            foreach (var (index, item) in _items.Index())
            {
                Console.WriteLine($"{index + 1} - {item}");
            }
            PrintedLines += _items.Count;
        }

        private void ClearOutput()
        {
            Console.Write($"\e[{PrintedLines}F"); // Перемещает курсор вверх на PrintedLines строк
            Console.Write("\e[0J"); // Очищает консоль от курсора до конца экрана
        }
    }
}

File: Project/UI/BaseItemPicker.cpp
#include "UI/BaseItemPicker.hpp"

#include <iostream>
#include <format>


BaseItemPicker::BaseItemPicker(const std::vector<std::string>& items) {
	m_items = items;
}

int BaseItemPicker::GetItemCount() {
	return m_items.size();
}

int BaseItemPicker::Pick() {
    m_printedLines = 0;
    PrintItems();

    int selectedIndex = GetInput();

    ClearOutput();
    std::cout << std::format("{} - {}\n", selectedIndex + 1, m_items[selectedIndex]);

    return selectedIndex;
}

void BaseItemPicker::PrintItems() {
    for (int i = 0; i < m_items.size(); i++) {
        std::cout << std::format("{} - {}\n", i + 1, m_items[i]);
    }
    m_printedLines += m_items.size();
}

void BaseItemPicker::ClearOutput() {
    std::cout << "\u001b[" << m_printedLines << "F"; // Перемещает курсор вверх на m_printedLines строк
    std::cout << "\u001b[0J"; // Очищает консоль от курсора до конца экрана
}
